File: debug_detailed_state.ts
#!/usr/bin/env bun
/**
 * Detailed Kalman state logging to find divergence point
 */

import { readFileSync } from "node:fs";
import { parse } from "csv-parse/sync";

// Import from typescript_lib
import {
  InMemoryStore,
  processMeasurement,
} from "./typescript_lib/src/index";

// Import service layer
import {
  type MeasurementInput,
} from "./services/weight_processor_service";

function getDefaultConfig(): any {
  return {
    database: {
      backend: "memory",
      table_name: "weight-processor-state",
      region: "us-east-1",
    },
    kalman: {
      initial_variance: 0.364,
      transition_covariance_weight: 0.018,
      transition_covariance_trend: 0.00012,
      observation_covariance: 5,
      reset: {
        initial: {
          enabled: true,
          initial_variance_multiplier: 10,
          weight_noise_multiplier: 50,
          trend_noise_multiplier: 50,
          observation_noise_multiplier: 20,
          adaptation_measurements: 10,
          adaptation_days: 10,
          adaptation_decay_rate: 2.5,
        },
        hard: {
          enabled: true,
          gap_threshold_days: 30,
          initial_variance_multiplier: 3,
          weight_noise_multiplier: 5,
          trend_noise_multiplier: 50,
          observation_noise_multiplier: 0.7,
          adaptation_measurements: 10,
          adaptation_days: 7,
          adaptation_decay_rate: 2.5,
        },
        soft: {
          enabled: true,
          min_weight_change_kg: 5,
          cooldown_days: 3,
          trigger_sources: ["questionnaire", "care-team-upload"],
          initial_variance_multiplier: 2,
          weight_noise_multiplier: 5,
          trend_noise_multiplier: 20,
          observation_noise_multiplier: 0.7,
          adaptation_measurements: 15,
          adaptation_days: 10,
          adaptation_decay_rate: 2.5,
        },
      },
    },
    quality_scoring: {
      use_harmonic_mean: true,
      threshold: 0.46,
      components: {
        kalman_fit: { weight: 0.4, enabled: true },
        temporal_consistency: { weight: 0.3, enabled: true },
        anomaly_detection: { weight: 0.2, enabled: true },
        source_reliability: { weight: 0.05, enabled: true },
        trend_alignment: { weight: 0.05, enabled: true },
      },
      component_weights: {
        kalman_fit: 0.4,
        temporal_consistency: 0.3,
        anomaly_detection: 0.2,
        source_reliability: 0.05,
        trend_alignment: 0.05,
      },
    },
    replay: {
      buffered_replay_enabled: false,  // Disable for clearer debugging
      buffer_hours: 24,
      max_buffer_measurements: 100,
    },
  };
}

async function main() {
  const csvFile = "/tmp/debug_user_full.csv";
  const userId = "ADC64C0B-CB46-41F9-BDA0-CC11A35942D7";

  console.log("=== TypeScript Detailed State Debug ===\n");

  // Load CSV
  const csvContent = readFileSync(csvFile, "utf-8");
  const rows = parse(csvContent, {
    columns: true,
    skip_empty_lines: true,
  });

  // Parse measurements
  const measurements: MeasurementInput[] = rows.map((row: any) => ({
    measurementId: row.id,
    weight: parseFloat(row.value_quantity),
    unit: row.unit || "kg",
    timestamp: new Date(row.effective_date_time || row.timestamp),
    source: row.source_type || "unknown",
  }));

  // Sort chronologically
  measurements.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());

  // Initialize storage and config
  const stateStore = new InMemoryStore();
  const config = getDefaultConfig();

  // Process measurements and log state for #47-50
  let measurementCount = 0;

  for (const measurement of measurements) {
    measurementCount++;

    // Get state BEFORE processing
    const stateBefore = await stateStore.getState(userId);

    const result = await processMeasurement(
      userId,
      measurement.weight,
      measurement.timestamp,
      measurement.source,
      config,
      measurement.unit,
      stateStore,
      null
    );

    // Get state AFTER processing
    const stateAfter = await stateStore.getState(userId);

    // Log detailed state for measurements 47-50
    if (measurementCount >= 47 && measurementCount <= 50) {
      console.log(`\n${"=".repeat(80)}`);
      console.log(`MEASUREMENT #${measurementCount}: ${measurement.measurementId}`);
      console.log(`${"=".repeat(80)}`);
      console.log(`Date: ${measurement.timestamp.toISOString()}`);
      console.log(`Weight: ${measurement.weight} kg`);
      console.log(`Source: ${measurement.source}`);

      if (stateBefore) {
        console.log(`\n--- STATE BEFORE ---`);
        if (stateBefore.kalman_params) {
          const kp = stateBefore.kalman_params;
          console.log(`Kalman State (x):`);
          console.log(`  Weight estimate: ${kp.state_mean?.[0] ?? 'N/A'}`);
          console.log(`  Trend (velocity): ${kp.state_mean?.[1] ?? 'N/A'}`);
          console.log(`Kalman Covariance (P):`);
          console.log(`  P[0,0]: ${kp.state_covariance?.[0]?.[0] ?? 'N/A'}`);
          console.log(`  P[0,1]: ${kp.state_covariance?.[0]?.[1] ?? 'N/A'}`);
          console.log(`  P[1,0]: ${kp.state_covariance?.[1]?.[0] ?? 'N/A'}`);
          console.log(`  P[1,1]: ${kp.state_covariance?.[1]?.[1] ?? 'N/A'}`);
          console.log(`Process Noise (Q):`);
          console.log(`  Q[0,0]: ${kp.process_noise_covariance?.[0]?.[0] ?? 'N/A'}`);
          console.log(`  Q[1,1]: ${kp.process_noise_covariance?.[1]?.[1] ?? 'N/A'}`);
          console.log(`Observation Noise (R): ${kp.observation_covariance ?? 'N/A'}`);
        }
        if (stateBefore.last_raw_weight) {
          console.log(`Last raw weight: ${stateBefore.last_raw_weight}`);
        }
        if (stateBefore.last_timestamp) {
          console.log(`Last timestamp: ${stateBefore.last_timestamp}`);
        }
      }

      console.log(`\n--- PROCESSING RESULT ---`);
      console.log(`Accepted: ${result.accepted}`);
      console.log(`Quality Score: ${result.quality_score?.toFixed(6) ?? 'N/A'}`);
      if (result.quality_components) {
        console.log(`Components:`);
        console.log(`  kalman_fit: ${result.quality_components.kalman_fit?.toFixed(6) ?? 'N/A'}`);
        console.log(`  temporal_consistency: ${result.quality_components.temporal_consistency?.toFixed(6) ?? 'N/A'}`);
        console.log(`  anomaly_detection: ${result.quality_components.anomaly_detection?.toFixed(6) ?? 'N/A'}`);
        console.log(`  source_reliability: ${result.quality_components.source_reliability?.toFixed(6) ?? 'N/A'}`);
        console.log(`  trend_alignment: ${result.quality_components.trend_alignment?.toFixed(6) ?? 'N/A'}`);
      }
      if (result.kalman_estimate !== undefined) {
        console.log(`Kalman estimate: ${result.kalman_estimate.toFixed(6)}`);
      }
      if (result.kalman_variance !== undefined) {
        console.log(`Kalman variance: ${result.kalman_variance.toFixed(6)}`);
      }
      if (result.trend !== undefined) {
        console.log(`Trend (velocity): ${result.trend.toFixed(8)}`);
      }
      if (result.innovation !== undefined) {
        console.log(`Innovation: ${result.innovation.toFixed(6)}`);
      }
      if (result.normalized_innovation !== undefined) {
        console.log(`Normalized innovation: ${result.normalized_innovation.toFixed(6)}`);
      }

      if (stateAfter) {
        console.log(`\n--- STATE AFTER ---`);
        if (stateAfter.kalman_params) {
          const kp = stateAfter.kalman_params;
          console.log(`Kalman State (x):`);
          console.log(`  Weight estimate: ${kp.state_mean?.[0] ?? 'N/A'}`);
          console.log(`  Trend (velocity): ${kp.state_mean?.[1] ?? 'N/A'}`);
          console.log(`Kalman Covariance (P):`);
          console.log(`  P[0,0]: ${kp.state_covariance?.[0]?.[0] ?? 'N/A'}`);
          console.log(`  P[0,1]: ${kp.state_covariance?.[0]?.[1] ?? 'N/A'}`);
          console.log(`  P[1,0]: ${kp.state_covariance?.[1]?.[0] ?? 'N/A'}`);
          console.log(`  P[1,1]: ${kp.state_covariance?.[1]?.[1] ?? 'N/A'}`);
        }
      }
    }
  }

  console.log(`\n\nProcessed all ${measurementCount} measurements`);
}

main().catch(console.error);

File: debug_detailed_state.py
#!/usr/bin/env python3
"""Detailed Kalman state logging to find divergence point"""

import csv
from datetime import datetime, timezone
from pathlib import Path
import sys

# Add paths for imports
project_root = Path(__file__).parent
sys.path.insert(0, str(project_root / "python_lib" / "src"))
sys.path.insert(0, str(project_root / "be_implementation_service" / "src"))

# Import from be_implementation_service
from aws.api.models import Measurement
from aws.config.config_manager import ConfigManager

# Import from python_lib
from weight_processor_lib.core.database import InMemoryStore
from weight_processor_lib.core.processing.processor import process_measurement


def parse_timestamp(date_str: str) -> datetime:
    """Parse timestamp string to datetime."""
    if not date_str:
        return datetime.now(timezone.utc)
    try:
        if "T" in date_str:
            dt = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
            if dt.tzinfo is None:
                return dt.replace(tzinfo=timezone.utc)
            return dt.astimezone(timezone.utc)
        elif " " in date_str:
            for fmt in ["%Y-%m-%d %H:%M:%S.%f", "%Y-%m-%d %H:%M:%S"]:
                try:
                    dt = datetime.strptime(date_str, fmt)
                    return dt.replace(tzinfo=timezone.utc)
                except ValueError:
                    continue
            raise ValueError(f"Cannot parse: {date_str}")
        else:
            dt = datetime.strptime(date_str, "%Y-%m-%d")
            return dt.replace(tzinfo=timezone.utc)
    except Exception:
        return datetime.now(timezone.utc)


def safe_format(val, fmt=".6f"):
    """Safely format a value or return N/A"""
    if val is None:
        return "N/A"
    try:
        return f"{val:{fmt}}"
    except:
        return str(val)


def main():
    csv_file = "/tmp/debug_user_full.csv"
    user_id = "ADC64C0B-CB46-41F9-BDA0-CC11A35942D7"

    print("=== Python Detailed State Debug ===\n")

    # Load CSV
    measurements = []
    with open(csv_file, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            measurement = Measurement(
                uuid=row.get("id"),
                weight=float(row.get("value_quantity", "0")),
                unit=row.get("unit", "kg"),
                effectiveDateTime=parse_timestamp(row.get("effective_date_time") or row.get("timestamp")),
                source=row.get("source_type", "unknown"),
            )
            measurements.append(measurement)

    # Sort chronologically
    measurements = sorted(measurements, key=lambda m: m.measured_at)

    # Initialize storage and config
    state_store = InMemoryStore()
    config = ConfigManager.load_config(source="file")
    config["database"]["backend"] = "memory"
    config["replay"]["buffered_replay_enabled"] = False  # Disable for clearer debugging

    # Process measurements and log state for #47-50
    measurement_count = 0

    for measurement in measurements:
        measurement_count += 1

        # Get state BEFORE processing
        state_before = state_store.get_state(user_id)

        result = process_measurement(
            user_id=user_id,
            weight=measurement.weight_value,
            timestamp=measurement.measured_at,
            source=measurement.source,
            config=config,
            unit=measurement.weight_unit,
            db=state_store,
            user_height_m=None
        )

        # Get state AFTER processing
        state_after = state_store.get_state(user_id)

        # Log detailed state for measurements 47-50
        if 47 <= measurement_count <= 50:
            print(f"\n{'='*80}")
            print(f"MEASUREMENT #{measurement_count}: {measurement.measurement_id}")
            print(f"{'='*80}")
            print(f"Date: {measurement.measured_at.isoformat()}")
            print(f"Weight: {measurement.weight_value} kg")
            print(f"Source: {measurement.source}")

            if state_before:
                print(f"\n--- STATE BEFORE ---")
                if "kalman_params" in state_before:
                    kp = state_before["kalman_params"]
                    print(f"Kalman State (x):")
                    state_mean = kp.get('state_mean', [None])
                    print(f"  Weight estimate: {safe_format(state_mean[0] if state_mean else None)}")
                    print(f"  Trend (velocity): {safe_format(state_mean[1] if len(state_mean) > 1 else None, '.8f')}")
                    print(f"Kalman Covariance (P):")
                    cov = kp.get('state_covariance', [[None]])
                    print(f"  P[0,0]: {safe_format(cov[0][0] if cov and len(cov) > 0 else None)}")
                    print(f"  P[0,1]: {safe_format(cov[0][1] if cov and len(cov) > 0 and len(cov[0]) > 1 else None)}")
                    print(f"  P[1,0]: {safe_format(cov[1][0] if cov and len(cov) > 1 else None)}")
                    print(f"  P[1,1]: {safe_format(cov[1][1] if cov and len(cov) > 1 and len(cov[1]) > 1 else None)}")
                    print(f"Process Noise (Q):")
                    q = kp.get('process_noise_covariance', [[None]])
                    print(f"  Q[0,0]: {safe_format(q[0][0] if q and len(q) > 0 else None)}")
                    print(f"  Q[1,1]: {safe_format(q[1][1] if q and len(q) > 1 and len(q[1]) > 1 else None, '.8f')}")
                    print(f"Observation Noise (R): {kp.get('observation_covariance', 'N/A')}")
                if "last_raw_weight" in state_before:
                    print(f"Last raw weight: {state_before['last_raw_weight']}")
                if "last_timestamp" in state_before:
                    print(f"Last timestamp: {state_before['last_timestamp']}")

            print(f"\n--- PROCESSING RESULT ---")
            print(f"Accepted: {result.get('accepted')}")
            print(f"Quality Score: {safe_format(result.get('quality_score'))}")
            
            components = result.get('quality_components', {})
            if components:
                print(f"Components:")
                for key in ['kalman_fit', 'temporal_consistency', 'anomaly_detection', 'source_reliability', 'trend_alignment']:
                    print(f"  {key}: {safe_format(components.get(key))}")
            
            print(f"Kalman estimate: {safe_format(result.get('kalman_estimate'))}")
            print(f"Kalman variance: {safe_format(result.get('kalman_variance'))}")
            print(f"Trend (velocity): {safe_format(result.get('trend'), '.8f')}")
            print(f"Innovation: {safe_format(result.get('innovation'))}")
            print(f"Normalized innovation: {safe_format(result.get('normalized_innovation'))}")

            if state_after:
                print(f"\n--- STATE AFTER ---")
                if "kalman_params" in state_after:
                    kp = state_after["kalman_params"]
                    print(f"Kalman State (x):")
                    state_mean = kp.get('state_mean', [None])
                    print(f"  Weight estimate: {safe_format(state_mean[0] if state_mean else None)}")
                    print(f"  Trend (velocity): {safe_format(state_mean[1] if len(state_mean) > 1 else None, '.8f')}")
                    print(f"Kalman Covariance (P):")
                    cov = kp.get('state_covariance', [[None]])
                    print(f"  P[0,0]: {safe_format(cov[0][0] if cov and len(cov) > 0 else None)}")
                    print(f"  P[0,1]: {safe_format(cov[0][1] if cov and len(cov) > 0 and len(cov[0]) > 1 else None)}")
                    print(f"  P[1,0]: {safe_format(cov[1][0] if cov and len(cov) > 1 else None)}")
                    print(f"  P[1,1]: {safe_format(cov[1][1] if cov and len(cov) > 1 and len(cov[1]) > 1 else None)}")

    print(f"\n\nProcessed all {measurement_count} measurements")


if __name__ == "__main__":
    main()
